Refund windows

bderusha opened the pull request:
## [ECOM-2595](https://openedx.atlassian.net/browse/ECOM-2595)

Only allow refunds either 14 days after purchase or 14 days after course start (whichever is later).

**Depends on https://github.com/edx/ecommerce/pull/412**

bderusha commented:
jenkins run all

bderusha commented:
@peter-fogg @clintonb Could you guys take a first pass look at this?

peter-fogg commented on openedx/core/djangoapps/commerce/__init__.py:
```diff
@@ -0,0 +1,36 @@
+""" Commerce app. """
```
Is the fact that we've got two Django apps named `commerce` going to cause trouble? You may also need to include this new app in the `INSTALLED_APPS` settings.

Though, actually, now that I think of it... Django 1.4 probably doesn't pick this up as an app at all, since it doesn't have a `models.py` file. So maybe this would be more appropriate to include in `openedx/core/lib`?

clintonb replied:
Yep, moving this to `openedx/core/lib/commerce.py` seems more reasonable.

peter-fogg commented on common/djangoapps/student/models.py:
```diff
@@ -1381,6 +1384,22 @@ def refundable(self):
+        order = ecommerce_api_client(self.user).orders(order_number).get()
+
+        refund_window_start_date = max(
+            datetime.strptime(order.get('date_placed', "1970-01-01T00:00:00Z"), "%Y-%m-%dT%H:%M:%SZ"),
```
Pull out `"1970-01-01T00:00:00Z"` into a named constant.

clintonb replied:
If the date_placed value is missing, there is an even bigger issue with our API or data. Don't provide a default. Definitely let the exception be raised.

clintonb replied:
The date-time format should be a constant provided by edx-rest-api-client. You'll need to add it.

peter-fogg commented on common/djangoapps/student/models.py:
```diff
@@ -1381,6 +1384,22 @@ def refundable(self):
+            name='order_number'
+        ).get('value')
+
+        order = ecommerce_api_client(self.user).orders(order_number).get()
```
What happens if there's no corresponding  `CourseEnrollmentAttribute` and `order_number` is `None`? Will an exception be thrown?

peter-fogg commented on common/djangoapps/student/migrations/0053_auto__add_enrollmentrefundconfiguration.py:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    complete_apps = ['student']
```
Missing newline.

bderusha replied:
No migration files have the extra new line.  They aren't generated with one. I'd prefer to not touch a generated file if we don't have to

peter-fogg replied:
Oh, right. Never mind then.

clintonb commented on common/djangoapps/student/models.py:
```diff
@@ -2024,3 +2043,63 @@ def get_enrollment_attributes(cls, enrollment):
+    # able to replace this field with a DurationField named
+    # `refund_window` without having to run a migration or change
+    # other code.
+    refund_window_microseconds = models.BigIntegerField(
```
We definitely don't need microsecond granularity. Days should be sufficient.

peter-fogg replied:
I definitely agree that the granularity is excessive, but I want to ensure that we can smoothly change this to a `DurationField` after the Django 1.8 upgrade. I'd rather use the officially-supported field types when we can, rather than rolling our own.

clintonb commented on common/djangoapps/student/models.py:
```diff
@@ -2024,3 +2043,63 @@ def get_enrollment_attributes(cls, enrollment):
             for attribute in cls.objects.filter(enrollment=enrollment)
         ]
+
+    @classmethod
```
Why not make this an instance method, and return the actual attribute?

bderusha commented:
@peter-fogg @clintonb I've implemented your previous recommendations and added test coverage.  I believe this is now ready for final review (pending passing build)

peter-fogg commented on common/djangoapps/student/models.py:
```diff
@@ -1374,7 +1375,9 @@ def refundable(self):
 
-        #TODO - When Course administrators to define a refund period for paid courses then refundable will be supported. # pylint: disable=fixme
+        # If it is after the refundable window they should not be refunded.
+        if self._refund_window_end_date() and datetime.now() > self._refund_window_end_date():
```
Don't call `self._refund_window_end_date` twice -- it's got multiple DB calls and a network call.

bderusha replied:
Good catch! Fixing

peter-fogg commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +226,124 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+
+        self.assertFalse(enrollment.refundable())
+
+    # WIP
```
Still WIP?

bderusha replied:
Nope. Fixed.

bderusha commented on openedx/core/lib/commerce.py:
```diff
@@ -0,0 +1,38 @@
+""" Commerce API Service. """
```
FYI @cpennington This is the ecommerce client pulled out of LMS.

cpennington replied:
It references django, so it should probably go into `openedx/core/djangoapps`

bderusha replied:
@cpennington moved. please have a look

peter-fogg commented:
Once those comments are addressed and the quality build is fixed, :+1:

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -2,17 +2,26 @@
 import ddt
+import httpretty
 import unittest
+from datetime import datetime, timedelta
```
Minor: alphabetize the import groups.

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
 
         return self.client.post(reverse('change_enrollment'), params)
+
+    def _enrollment(self):
```
Make the name describe what the method does: `create_enrollment`.

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+
+    def _enrollment(self):
+        """ Generate a default course mode and enrollment. """
+        mode = CourseMode(
```
A few issues:
1. Why are you creating a course mode here here?
2. Use `CourseMode.objects.create()`.
3. Does this belong in `setUp()`?

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+        expected_date = now + expected_date_delta
+        window_days = timedelta(days=days)
+
+        order_number = 'OSCR1000'
```
Nit: `OSCR-1000`

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+            adding_headers={'Content-Type': JSON}
+        )
+
+        enrollment = CourseEnrollment(
```
Use `CourseEnrollment.objects.create()` to instantiate and save in one call.

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+            self.assertTrue(enrollment.refundable())
+
+    def test_refundable_no_verified(self):
+        """ Assert that enrollment is not refundable if there is no verified course mode for the course."""
```
Why does the verified mode matter? Shouldn't this be based on whether the user paid?

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+
+    def test_refundable_can_refund(self):
+        """ Assert that enrollment is refundable if can_refund is set."""
+        enrollment = self._enrollment()
```
This is called for nearly every test. Put it in `setUp()`.

clintonb commented on common/djangoapps/student/tests/test_enrollment.py:
```diff
@@ -218,3 +227,126 @@ def _change_enrollment(self, action, course_id=None, email_opt_in=None):
+        )
+
+    def test_refundable_can_refund(self):
+        """ Assert that enrollment is refundable if can_refund is set."""
```
I see the current code has `can_refund` set by some external logic. Can we move that logic into the `refundable()` method and eliminate this pseudo-property?

bderusha replied:
I agree that `can_refund` is a hack that should be resolved.  I think it is out of scope for this ticket.  I think we should tackle this when we pull the refund logic into Otto.

bderusha commented:
jenkins run python

bderusha commented:
@clintonb @cpennington addressed your comments

clintonb commented on __init__.py:
```diff
@@ -0,0 +1 @@
+__author__ = 'williamderusha'
```
Is this file necessary?

bderusha replied:
PyCharm auto generated file.  Not sure how that go there.  Will remove

bderusha replied:
Fixed

clintonb commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -404,6 +414,91 @@ def test_refundable_when_certificate_exists(self):
         self.assertFalse(enrollment.refundable())
 
     @unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+    def test_refundable_when_refund_window(self):
```
Add a docstring explaining what this test does.

bderusha replied:
Fixed.

clintonb commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -404,6 +414,91 @@ def test_refundable_when_certificate_exists(self):
 
     @unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+    def test_refundable_when_refund_window(self):
+        CourseModeFactory.create(
```
This is duplicated three times. Move it to `setUp`. Also, I just noticed that this file is HUGE! At a minimum, break the refund tests into their own file.

clintonb replied:
You might also consider creating a `VerifiedCourseModeFactory`.

bderusha replied:
Fixed. Moved to new file and put repeated code in `setup`

clintonb commented on common/djangoapps/student/models.py:
```diff
@@ -1382,6 +1386,22 @@ def refundable(self):
         else:
             return True
 
+    def _refund_window_end_date(self):
```
I just realized why I've found this so confusing. What do you think of using the phrase "refund cutoff" instead of window? Window implies a start date as well. While the order placement date is the start date, it's not actually used.

clintonb replied:
Also, you might as well make this a public method. I can see us wanting to display this value in the future. You'll also be able to appease PyLint without disabling checks.

bderusha replied:
Yes!  Much better name.  Updated and made public.

clintonb commented on common/djangoapps/student/models.py:
```diff
@@ -1382,6 +1386,22 @@ def refundable(self):
+        order_number = attribute.value
+        order = ecommerce_api_client(self.user).orders(order_number).get()
+        refund_window_start_date = max(
+            datetime.strptime(order.get('date_placed'), ECOMMERCE_DATE_FORMAT),
```
Use `order['date_placed']` so that the code fails more immediately (with a `KeyError` instead of `TypeError`) if the API returns invalid data.

bderusha replied:
Fixed.

clintonb commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -404,6 +414,91 @@ def test_refundable_when_certificate_exists(self):
+        self.assertTrue(enrollment.refundable())
+
+        with patch('student.models.CourseEnrollment._refund_window_end_date') as window:
+            window.return_value = datetime.now() - timedelta(days=1)
```
Add comments explaining what you're testing. It may seem obvious to you now; but, it won't be so obvious in a few weeks.

bderusha replied:
Fixed.

clintonb commented on common/djangoapps/student/models.py:
```diff
@@ -1382,6 +1386,22 @@ def refundable(self):
+        order = ecommerce_api_client(self.user).orders(order_number).get()
+        refund_window_start_date = max(
+            datetime.strptime(order.get('date_placed'), ECOMMERCE_DATE_FORMAT),
+            self.course_overview.start.replace(tzinfo=None)
```
Shouldn't this be in UTC timezone?

bderusha replied:
Not sure where I got this from.  Updated to be UTC

bderusha replied:
I remember now.  This was done in an attempt to make the two dates comparable. Right now you get the following error when you don't make `course_overview.start` naive 

`TypeError: can't compare offset-naive and offset-aware datetimes`

clintonb commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -404,6 +414,91 @@ def test_refundable_when_certificate_exists(self):
+    @override_settings(ECOMMERCE_API_SIGNING_KEY=TEST_API_SIGNING_KEY, ECOMMERCE_API_URL=TEST_API_URL)
+    @unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+    def test_refund_window_end_date(self, order_date_delta, course_start_delta, expected_date_delta, days):
+        """ Assert that the later refund date is and config window timedelta are used to calculate window end date."""
```
If I understand this correctly, you only want to verify that the cutoff/window end date is the maximum of either the order date or the course start date. If that's the case you only need two tests. Why use four?

bderusha replied:
We are also testing that it does the right thing for multiple refund periods.  The first two test that we get the expected result using the default 14 days. The last two use a period of 1 day to ensure that we get the correct results when using a non-default

bderusha replied:
I could break this into two separate tests, but for what it is it doesn't seem worth it.  Open to a second opinion on that though

clintonb commented on common/djangoapps/student/tests/tests.py:
```diff
@@ -404,6 +414,91 @@ def test_refundable_when_certificate_exists(self):
+    @ddt.unpack
+    @httpretty.activate
+    @override_settings(ECOMMERCE_API_SIGNING_KEY=TEST_API_SIGNING_KEY, ECOMMERCE_API_URL=TEST_API_URL)
+    @unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
```
Can this be applied to the entire test case? Does it need to be on every test method?

clintonb commented:
@bderusha Everything looks good. As you can see most of my concerns are around naming, comments and code organization.

bderusha commented:
@clintonb thanks for all the feedback. I think I've addressed all of your comments. Either with code fixes or explanations/discussion. Let me know what you think.

clintonb commented on common/djangoapps/student/tests/test_refunds.py:
```diff
@@ -0,0 +1,163 @@
+
+@ddt.ddt
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Test only valid in lms')
+class RefundableTest(ModuleStoreTestCase):
```
Use `SharedModuleStoreTestCase`. It should be faster.

clintonb commented on common/djangoapps/student/tests/test_refunds.py:
```diff
@@ -0,0 +1,163 @@
+    @httpretty.activate
+    @override_settings(ECOMMERCE_API_SIGNING_KEY=TEST_API_SIGNING_KEY, ECOMMERCE_API_URL=TEST_API_URL)
+    def test_refund_cutoff_date(self, order_date_delta, course_start_delta, expected_date_delta, days):
+        """ Assert that the later refund date is and config window timedelta are used to calculate window end date."""
```
Grammatical error.

clintonb commented:
:+1:

bderusha commented:
jenkins run bokchoy

bderusha commented:
jenkins run bokchoy
